### PR TITLE
PorousFlow: initial version of capillary pressure user objects

### PR DESCRIPTION
**Pull request by cpgr**
Following discussion in #9105, this is the initial implementation of capillary pressure as a user object rather than having a separate material for each formulation.

The benefits of this is being able to only enter capillary pressure values in one place in the input, rather than in several. This will also eventually allow us to deprecate and remove a bunch of materials that have the capillary pressures hard coded in, reducing the code base in PorousFlow.

Only the constant and van Genuchten forms have been added yet, and only for the PorousFlowFluidState materials. Eventually, the remaining materials can be migrated over to this design. I just didn't want to change 200 tests right now.

Refs #9105

### CODE REVIEW

**Comment by cpgr** [comment]
I should add, I fixed up an error with some derivatives that became apparent using the van Genuchten capillary pressures.

**Comment by cpgr** [comment]
@WilkAndy Any thoughts?

**Review by permcody — APPROVED** [review]

File: modules/porous_flow/src/userobjects/PorousFlowCapillaryPressure.C [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+template <>
+InputParameters
+validParams<PorousFlowCapillaryPressure>()
+{
+  InputParameters params = validParams<GeneralUserObject>();
+  params.addRangeCheckedParam<Real>(
+      "sat_lr",
+      0.0,
+      "sat_lr >= 0 & sat_lr <= 1",
+      "Liquid residual saturation.  Must be between 0 and 1. Default is 0");
```
Comment (WilkAndy) [inline modules/porous_flow/src/userobjects/PorousFlowCapillaryPressure.C]: Sorry for not seeing this PR.  What is this sat_lr, by the way?  Is it the same as might enter the relative permeability, or can it be different?

File: modules/porous_flow/src/userobjects/PorousFlowCapillaryPressureVG.C [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    _alpha(getParam<Real>("alpha")),
+    _p0(1.0 / _alpha),
+    _pc_max(getParam<Real>("pc_max"))
+{
+}
+
+Real
+PorousFlowCapillaryPressureVG::capillaryPressure(Real saturation) const
+{
+  Real seff = effectiveSaturationFromSaturation(saturation);
```
Comment (WilkAndy) [inline modules/porous_flow/src/userobjects/PorousFlowCapillaryPressureVG.C]: these "Real seff" could be "const Real seff"

**Review by WilkAndy — COMMENTED** [review]
Thanks @cpgr.   Late comments on this merged PR!

File: modules/porous_flow/src/userobjects/PorousFlowCapillaryPressure.C (hunk 1 above)
Comment (cpgr) [inline modules/porous_flow/src/userobjects/PorousFlowCapillaryPressure.C]: It is used in TOUGH2 where they recommend setting it slightly smaller than the rel perm value to aid convergence. I added it so I can properly match the capillary pressure used in TOUGH2 when comparing results.

File: modules/porous_flow/src/userobjects/PorousFlowCapillaryPressureVG.C (hunk 2 above)
Comment (cpgr) [inline modules/porous_flow/src/userobjects/PorousFlowCapillaryPressureVG.C]: I'll fix this in the next associated PR. Thanks

File: modules/porous_flow/src/userobjects/PorousFlowCapillaryPressure.C (hunk 1 above)
Comment (WilkAndy) [inline modules/porous_flow/src/userobjects/PorousFlowCapillaryPressure.C]: Aha, i get it.  i've never put sat_lr = residual_from_relperm: i've always used sat_lr = 0.  Similarly in my old COSFLOW code we had an "immobile saturation" (which is the residual saturation in the relperm) and a "residual saturation" (which is this sat_lr) and we ran lots of rather large models with them both zero.   i wonder whether we would have gained better convergence if we used TOUGH's convention.

When they say "slightly smaller", do they mean sat_lr=0.999*residual_in_relperm, or sat_lr=0.1*residual_in_relperm, or something else?   What do you usually do?   I'd really like to speed up models by changing these parameters, considering that they're never actually measured accurately (for entire rock masses, especially!)

**Comment by WilkAndy** [comment]
When are you planning to do the remaining Materials, by the way?  I'm on holidays for the next week so am not using MOOSE during that time.... hint hint!

**Comment by cpgr** [comment]
Ha ha! Enjoy your break. I'll try and get onto it before you are back.

**Pull request merged**